apps/05Mastermind.js

```diff
@@ -4,75 +4,115 @@ var assert = require('assert');
 var solution = '';
+// var solution = 'input the answer here!';
 var letters = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h'];
 
 function printBoard() {
-  for (var i = 0; i < board.length; i++) {
-    console.log(board[i]);
-  }
+    for (var i = 0; i < board.length; i++) {
+        console.log(board[i]);
+    }
 }
 
-function generateSolution() {
-  for (var i = 0; i < 4; i++) {
-    var randomIndex = getRandomInt(0, letters.length);
-    solution += letters[randomIndex];
-  }
+function generateSolution(solution, guess) {
+    for (var i = 0; i < 4; i++) {
+        var randomIndex = getRandomInt(0,
+            letters.length);
+        solution += letters[randomIndex];
+    }
 }
 
 function getRandomInt(min, max) {
-  return Math.floor(Math.random() * (max - min)) + min;
+    return Math.floor(Math.random() * (max - min)) + min;
 }
 
-function generateHint() {
-  // your code here
+function generateHint(solution, guess) {
+    // your code here
+    var correctLetterLocation = 0;
+    // (a counter always starts at zero)
+    var correctLetters = 0;
+    // following is split, which separates a string into an array:
+    var solutionArray = solution.split('');
+    var guessArray = guess.split('');
+
+    // so, if true for generate Hint:
+    // if (solutionArray[0] === guessArray[0] {
+    //   correctLetterLocations ++;})
+    // But instead, combine the ifs into one:
+
+    for (var i = 0; i < solutionArray.length; i++) {
+        if (solutionArray[i] === guessArray[i]) {
+            correctLetterLocation++;
+            //  The above is a very typical code.
+            // Below is giving a null value to the letter already checked:
+            solutionArray[i] = null;
+        }
+    }
+
+    for (var i = 0; i < guessArray.length; i++) {
+        var targetIndex = solutionArray.indexOf(guessArray[i]);
+        if (targetIndex > -1) {
+            correctLetters++;
+            solutionArray[targetIndex] = null;
+        }
+    } return correctLetterLocation + "-" + correctLetters;
 }
 
 function mastermind(guess) {
-  // your code here
-}
+    // your code here
+    if (guess === solution) {
+        return "You guessed it!";
+    }
+    var hint = generateHint(solution, guess);
+    board.push(guess + "" + hint);
+
+    if (board.length === 10) {
+        return "You ran out of turns! The solution was " + solution;
+    }
+    return guess;
 
+}
 
 function getPrompt() {
-  rl.question('guess: ', (guess) => {
-    console.log( mastermind(guess) );
-    printBoard();
-    getPrompt();
-  });
+    rl.question('select four letters a-h: ', (guess) => {
+        console.log(mastermind(guess));
+        printBoard();
+        getPrompt();
+    });
 }
 
 // Tests
 
-if (typeof describe === 'function') {
-
-  describe('#mastermind()', function () {
-    it('should register a guess and generate hints', function () {
-      solution = 'abcd';
-      mastermind('aabb');
-      assert.equal(board.length, 1);
+if (typeof describe === 'function ') {
```

Comment:
Not sure how this happened, but there is an extra space in there 'function ' which is making the tests not run